### ISSUE TITLE
feat(beam): zero-allocation beam search engine spine with flat arena DAG and O(B*K) candidate expansion

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12407]
# feat(beam): zero-allocation beam search engine spine with flat arena DAG and O(B*K) candidate expansion

```routing
lane: internal/beam
paths: internal/beam/beam.go, internal/beam/pool.go, internal/beam/arena.go, internal/beam/beam_test.go
expected_steps: 4
```

**Lane:** internal/beam  
**Work Unit:** leaf (spine)  
**Generation:** gen/now  
**Priority:** P1  

## Parent context
Part of the Native Inference & Serving Tier (`internal/beam`, Tier 3 Mechanism), establishing multi-hypothesis beam search in `fak`'s zero-external-dependency Go engine for deterministic structured extraction and AST generation.

## Current state
`fak`'s in-kernel reference engine (`internal/model/kv.go` and `internal/modelengine/modelengine.go`) supports only greedy argmax decoding (`argmaxF32`), temperature/top-p nucleus sampling, and speculative candidate verification. There is no beam search or multi-hypothesis search capability in the token generation path. Naive beam search implementations in Go suffer from severe GC pressure (allocating token slices and candidate maps per step) and $O(B \cdot |\mathcal{V}|)$ sorting overhead ($>600{,}000$ floats per step for Qwen3.8).

## Why this is next
Deterministic extraction, strict JSON-schema compliance, AST compilation, and formal verification tasks require exploring multiple high-likelihood sequence hypotheses to avoid dead ends. A zero-allocation beam search spine provides the foundation for high-quality structured generation without memory bus saturation or GC pauses.

## Working spine
Pre-allocate flat `BeamNode` arena DAG -> extract local top-$2B$ candidate tokens per beam -> pool and select global top-$B$ survivors in $<5\,\mu\text{s}$ -> apply Wu et al. length normalization and theoretical upper-bound early stopping -> deliver verified beam search with zero heap allocations per step.

## Core through-line
1. Create package `internal/beam` registered at Tier 3 (Mechanism) in `internal/architest/architest_test.go`, cleanly importing `internal/abi` and `internal/model`.
2. Implement `BeamNode` and `BeamPool` in `internal/beam/pool.go` using a pre-allocated flat arena DAG with `ParentIndex` backpointers, eliminating per-step slice allocations.
3. Implement $O(B \cdot K)$ candidate filtering: extract local top-$2B$ candidate tokens from each beam using an in-place partial selection sort, merge into a $2B^2$ pool, and extract global top-$B$ survivors.
4. Implement Wu et al. / Google NMT length normalization ($lp(Y) = (5 + |Y|)^\alpha / 6^\alpha$) and theoretical upper-bound early stopping ($\text{UB}_t(b) = s_t^{(b)} / lp(t)$).
5. Author comprehensive unit tests in `internal/beam/beam_test.go` with synthetic model logits verifying `0 allocs/op` during iterative expansion and exact hypothesis ranking.

## In scope
- Tier 3 leaf package `internal/beam`.
- Flat arena DAG `BeamNode` with parent-pointer backtracking.
- Zero-allocation `BeamPool` supporting beam width $B \in [1, 16]$.
- $O(B \cdot K)$ top-$2B$ candidate filtering and selection.
- Wu et al. length normalization with configurable length penalty $\alpha \in [0.0, 2.0]$.
- Theoretical upper-bound early stopping heuristic.
- Benchmarks verifying zero heap allocations per step (`BenchmarkBeamStepAllocations`).

## Out of scope
- Multi-GPU distributed beam synchronization.
- Complex grammar parsing (handled via `DecodeConstraint` in follow-on).
- Multi-turn agent tool execution during search.

## Gold-plating boundary
Strictly excluded:
- Dynamic tree visualization UI or terminal renderers.
- External CGo / Python library bindings.

## Verifiable Witness
```bash
go test -v ./internal/beam -run TestBeamSearchSpine
go test -bench=BenchmarkBeamStep -benchmem ./internal/beam
```

## Done condition
`internal/beam` executes multi-hypothesis beam search over synthetic logits, reconstructing exact top-$B$ token sequences via backpointers, satisfying length penalties and early stopping, with `0 B/op` and `0 allocs/op` on the decode step.

## Scoped Acceptance Criteria
- [ ] 1. `internal/beam` is registered at Tier 3 in `internal/architest/architest_test.go` and passes `TestNoUpwardImports`. `[SW-VERIFIED]`
- [ ] 2. `BeamPool` executes $B=4$ expansion over $L=64$ steps with `0 allocs/op` in benchmark. `[SW-VERIFIED]`
- [ ] 3. Top-$2B$ candidate pruning produces exact top-$B$ hypotheses matching full $B \times |\mathcal{V}|$ sort. `[SW-VERIFIED]`
- [ ] 4. Length normalization correctly breaks short-sequence bias compared to unnormalized log-probs. `[SW-VERIFIED]`

## Acceptance gate
`go test -v ./internal/beam/...` and `go test -v ./internal/architest -run TestNoUpwardImports` exit 0 with 0 failures.

## Closure binding
Close via an explicit-path trunk commit landing `internal/beam/` and its registration in `internal/architest/architest_test.go`.

## Likely files
- `internal/beam/beam.go`
- `internal/beam/pool.go`
- `internal/beam/arena.go`
- `internal/beam/beam_test.go`
- `internal/architest/architest_test.go`
